packages/compiler/lib/src/compilation/ast_hir_lowering/body.dart

```diff
@@ -673,6 +685,53 @@ class ExpressionContext extends InnerContext {
   }
 }
 
+class LoopContext extends InnerContext {
+  LoopContext(
+    Context parent,
+    this.id,
+    this.label,
+  )   : assert(id != null),
+        assert(label != null),
+        super(parent);
+
+  final DeclarationLocalId id;
+  final Option<String> label;
+  final _identifiers = <String, hir.Identifier>{};
+
+  @override
+  void addIdentifier(hir.LocalPropertyIdentifier identifier) {
+    _identifiers[identifier.name] = identifier;
+  }
+
+  @override
+  List<Identifier> resolveIdentifier(String name) {
+    final result = _identifiers[name];
+    if (result != null) return [result];
+    return parent.value.resolveIdentifier(name);
+  }
+
+  @override
+  Option<Tuple2<DeclarationLocalId, Option<hir.CandyType>>> resolveBreak(
+    Option<String> label,
+  ) {
+    if (label is None ||
+        label == this.label ||
+        this.label is None && label == Some('loop')) {
+      return Some(Tuple2(id, parent.value.expressionType));
+    }
+    return parent.flatMapValue((context) => context.resolveBreak(label));
+  }
+}
+
+class BreakExpressionVisitor extends DoNothingExpressionVisitor {
+  final breakTypes = <hir.CandyType>{};
+
+  @override
+  void visitBreakExpression(BreakExpression node) {
+    breakTypes.add(node.expression.type);
```

Comment:
`expression` is nullable
```suggestion
    breakTypes.add(node.expression?.type ?? hir.CandyType.unit);
```